### PR TITLE
Port to TypeScript + 0.10.0 Gamemode Changes

### PR DESCRIPTION
**Pull request by tsa96**
Working on Pano seriously for the first time in ages doing gamemode changes and was very unhappy with the quality of the TypeScript integration, mainly the namespacing issues caused by not being able to use imports. I eventually snapped and adding module loading support in C++, then found I was now *liking* the TypeScript integration so much I ended up porting everything across.

There were enough bugs, random unused events/APIs and other issues I encountered along the way that it was definitely justified. Would've preferred doing after 0.10.0 but hey, it meant we can get the shared types/enums from the website repo hooked up whilst doing all the updated timer, comparisons etc.

Momentum-specific types have been split out to this repo as discussed. I've almost upgraded all our deps, had to mess around with eslint in the process.

The one thing I haven't quite finished yet is the map selector. New modes are hooked up but I want to get zone downloading working by Saturday, maybe even do the submission maps tab if I have time. Everything else is good for review, if people reach the wip map selector stuff fast then I'll split that work out into a separate PR, but I expect it'll take at least a week for this work + all my other stuff to get reviewed.

Needs https://github.com/StrataSource/pano-typed/pull/22 on the shared types repo, then on red, my module loading PR and gamemode enum PR.

### Checks

-   [x] **I have thoroughly tested all of the code I have modified/added/removed to ensure something else did not break**
-   [x] I have followed [semantic commit messages](https://gist.github.com/joshbuchea/6f47e86d2510bce28f8e7f42ae84c716) e.g. `feat: Add foo`, `chore: Update bar`, etc...
-   [x] My branch has a clear history of changes that can be easy to follow when being reviewed commit-by-commit
-   [x] My branch is functionally complete; the only changes to be done will be those potentially requested in code review
-   [x] Fully tokenized all my strings (no hardcoded English strings!!) and supplied [bulk JSON strings](https://docs.momentum-mod.org/guide/localization/#bulk-adding-terms) below
-   [x] All changes requested in review have been `fixup`ed into my original commits.

### CODE REVIEW

**Comment by tsa96** [comment]
Will resolve conflicts with Peen's stuff once https://github.com/momentum-mod/panorama/pull/145 is in (and I've finally slept)

File: scripts/pages/settings/page.ts
Comment (tsa96) [inline scripts/pages/settings/page.ts:46]: Nah just get rid of traverseChildren stuff too complex a refactor for this pr

File: scripts/pages/settings/settings.ts
Comment (tsa96) [inline scripts/pages/settings/settings.ts:255]: Double check this works

File: scripts/pages/zoning/zoning.ts
Comment (tsa96) [inline scripts/pages/zoning/zoning.ts:1]: Merge latest Peen branch

File: scripts/util/toast-manager.ts
Comment (tsa96) [inline scripts/util/toast-manager.ts:173]: eh we use null quite a bit, do != null

File: scripts/hud/speedometer.ts
Comment (tsa96) [inline scripts/hud/speedometer.ts:4]: Let's move to new timer today

File: scripts/pages/settings/speedometer.ts
Comment (tsa96) [inline scripts/pages/settings/speedometer.ts:425]: `[255, 255, 255, 1]`?

File: scripts/util/module-helpers.ts
Comment (tsa96) [inline scripts/util/module-helpers.ts:72]: ".mts", grep for these and rename all

**Review by tsa96 — COMMENTED** [review]
Reviewing my own work since this is such a chore for everyone else to do, plus afaik I'm the only Pano person who works with TypeScript frequently.

Others reviewing is by-all-means appreciated, probably just need one other though this is such a slog to review.

I've tested pretty well ingame so far, going to do a second pass of ingame testing today as well.

File: scripts/hud/jump-stats.ts
Comment (tsa96) [reply scripts/hud/jump-stats.ts:144]: No

File: scripts/pages/leaderboards/entry.ts
Comment (tsa96) [reply scripts/pages/leaderboards/entry.ts:41]: nah just improve in future

File: scripts/pages/main-menu/main-menu.ts
Comment (tsa96) [reply scripts/pages/main-menu/main-menu.ts:113]: Getting mixed up with settings pages

File: scripts/pages/settings/speedometer.ts
Comment (tsa96) [reply scripts/pages/settings/speedometer.ts:425]: Nope, alpha is 255 for some reason

**Comment by tsa96** [comment]
Still need to get all the new timer stuff hooked up. Panzer's timer changes don't work on this branch (aren't in TypeScript with new data structures), currently my stuff is just overwriting his recent commit. 

I'll join them together properly soon, want to remove the old APIs in the process. Will also merge with Peen's new zoning stuff, just don't have energy tonight.

**Review by Gocnak — APPROVED** [review]
Split out the map selector stuff but the rest is good 👍

**Comment by tsa96** [comment]
Release tonight will be pretty janky anyway so no need to rush this in, we can just cut release off this branch without merging.

File: scripts/hud/cgaz.ts
Comment (PeenScreeker) [inline scripts/hud/cgaz.ts:739]: this is an extra `sqrt` call

File: scripts/hud/cgaz.ts
Comment (PeenScreeker) [inline scripts/hud/cgaz.ts:844]: just for my own interest, is this better than `Boolean(this.compassStatMode % 2)`?

File: scripts/hud/cgaz.ts
Comment (PeenScreeker) [inline scripts/hud/cgaz.ts:933]: should this explicitly return `number[]`?

Kind of related - what's the criteria for giving an explicit return type vs not?

File: scripts/hud/ground-boost.ts
Comment (PeenScreeker) [inline scripts/hud/ground-boost.ts:55]: for uniformity, should these config events be included in gamemode registration everywhere?

File: scripts/hud/jump-stats.ts
Comment (PeenScreeker) [inline scripts/hud/jump-stats.ts:35]: should this be listed in `events` when registering the panel for bhop?

File: scripts/hud/jump-stats.ts
Comment (PeenScreeker) [inline scripts/hud/jump-stats.ts:97]: I would rather add `.toFixed(0)` to `jumpCount` and `strafeCount` on lines 56, 69, and 74 and disallow `value` from being a `number`. I think `.toString()` on a `string` will be really cheap in v8 - this just feels extra to me. What do you think?

File: scripts/hud/jump-stats.ts
Comment (tsa96) [reply scripts/hud/jump-stats.ts:35]: Don't think so, as presumably we want to handle changes even if they happen when playing other modes. Or does the entire config get loaded from scratch when loading into a map?

File: scripts/hud/cgaz.ts
Comment (tsa96) [reply scripts/hud/cgaz.ts:739]: Oops!! Good spot. I thought it just find/replaced all of these, this is concerning. I'll go through and double check everything.

File: scripts/util/math.ts
Comment (PeenScreeker) [inline scripts/util/math.ts:24]: should this call the 2D or 3D version passing `vec` instead of redefining the operation?

File: scripts/hud/cgaz.ts
Comment (tsa96) [reply scripts/hud/cgaz.ts:844]: Performance-wise I don't know, though I suspect the direct comparison is faster. I find the code much clearer doing the comparison, since modulo checks like this are so common in every language, and doesn't require reasoning about truthiness of different numbers.

File: scripts/hud/synchronizer.ts
Comment (PeenScreeker) [inline scripts/hud/synchronizer.ts:88]: same question about registering events for game mode with the panel

File: scripts/hud/cgaz.ts
Comment (tsa96) [reply scripts/hud/cgaz.ts:933]: Good question. I'd say I'm general it's always best to add a return type, besides perhaps functions return `void` /async functions `Promise<void>`. Though when you put me on the spot and ask, I don't really see a good reason not to set a return type in that case, especially since `void` has an important *semantic* implication, i.e. that a function returning could technically return any value, but the caller must not *use* that value.

In this case, I was just being lazy and forgot to add a return type, it definitely seems helpful to be explicit in this case.

I suppose here's a cass where you might not want to specify a return type; TODO FINISH ON LAPOTP I'M NOT WRITING THIS SHIT ON PHONE KB

File: scripts/hud/jump-stats.ts
Comment (PeenScreeker) [reply scripts/hud/jump-stats.ts:35]: I think in every case where we would be registering a panel for a specific mode, the config settings get loaded as a part of `OnMapInit()`

File: scripts/hud/ground-boost.ts
Comment (tsa96) [reply scripts/hud/ground-boost.ts:55]: Oh, I think this is wrong and should be outside of the gamemode-specific stuff, so updates still happen if not in the given modes. Unless configs are always fetched on map load, will look at that when on my PC

File: scripts/hud/jump-stats.ts
Comment (tsa96) [reply scripts/hud/jump-stats.ts:35]: Good to know, I'll move everything in this case and leave a comment clarifying

File: scripts/hud/jump-stats.ts
Comment (tsa96) [reply scripts/hud/jump-stats.ts:97]: Gotcha, that seems a lot better. I was struggling to understand the types in this class, this makes a lot a sense.

File: scripts/util/math.ts
Comment (tsa96) [reply scripts/util/math.ts:24]: I guess so, functions calls in here probably get inlined so no performance benefit doing it this way.

File: scripts/hud/synchronizer.ts
Comment (tsa96) [reply scripts/hud/synchronizer.ts:88]: Yeah same response, if config is always loaded on map init we can pull all events into the famcy thing, will do that.

File: scripts/util/math.ts
Comment (tsa96) [reply scripts/util/math.ts:24]: Spent 5 mins playing around with this, indeed there's no noticeable performance difference

**Comment by tsa96** [comment]
Okay I've addressed everything, good to merge once @PeenScreeker approves

File: scripts/util/math.ts
Comment (PeenScreeker) [inline scripts/util/math.ts:5]: I missed this on first review - sorry! We should be able to get the 2D magnitude of 3D vectors (XY size). The function stays the same, but `vec` should be allowed to be `vec3 | vec2`. CGAZ needs this functionality.

File: scripts/util/math.ts
Comment (PeenScreeker) [inline scripts/util/math.ts:15]: same here, `vec` should be allowed to be `vec3 | vec2`. In the generic case, the 3D magnitude of a `vec3` will be returned, but if `magnitude2D()` is called on a `vec3` explicitly, the XY-projected magnitude should be returned.

File: scripts/util/math.ts
Comment (PeenScreeker) [reply scripts/util/math.ts:24]: My bad, I wasn't super clear about this:
`return 'z' in vec ? sumOfSquares3D(vec) : sumOfSquares2D(vec);`

**Review by PeenScreeker — CHANGES_REQUESTED** [review]

**Review by PeenScreeker — APPROVED** [review]

**Pull request merged**
